### PR TITLE
Move cs_vSend into SocketSendData and resolve RecordBytesSent lock inconsistency

### PR DESCRIPTION
**Pull request by troygiorshev**
As it stands, `RecordBytesSent` is called twice, once under the `cs_vSend` lock, and once not.  This PR moves both calls out of the lock.

In doing this, I've made a change to SocketSendData.  Before, it required the caller to lock `cs_vSend`, now it locks `cs_vSend` itself.  The only case where this makes a difference is when performing an optimistic write (where MessageHandler is the thread that calls SocketSendData, not SocketHandler).  As justified in the commit message, I don't think this will cause any problems, and subjectively I think this is an improvement to how we're using the locks.  I'm more than happy to be told that I'm wrong :)

Additionally this PR brings PushMessage into line with the style guide, and does verious other cleanups.

This PR was originally related to #18784 but it now stands on its own.

### CODE REVIEW

File: src/net.cpp [hunk 1]
```diff
@@ -2813,35 +2811,34 @@ bool CConnman::NodeFullyConnected(const CNode* pnode)
...
-        pnode->nSendSize += nTotalSize;
+        pnode->mapSendBytesPerMsgCmd[msg.m_type] += total_size;
+        pnode->nSendSize += total_size;
 
         if (pnode->nSendSize > nSendBufferMaxSize)
             pnode->fPauseSend = true;
-        pnode->vSendMsg.push_back(std::move(serializedHeader));
-        if (nMessageSize)
+        pnode->vSendMsg.push_back(std::move(serialized_header));
+        if (message_size)
```
Comment (jnewbery) [inline src/net.cpp:2834]: Since you're cleaning everything up here, this if block (and the one above) should have braces added.

**Review by jnewbery — COMMENTED** [review]
Concept ACK.

It seems to me that there are a few additional members of `CNode` that should be guarded by the `cs_vSend` lock:

- `mapSendBytesPerMsgCmd`
- `nSendSize`
- `nSendOffset`

I don't think we want any of those to be accessed/updated while another thread is locking `cs_vSend` and updating `vSendMsg` and `nSendBytes`. If you agree, what do you think about updating them in this PR?

File: src/net.cpp (hunk 1 above)
Comment (troygiorshev) [reply src/net.cpp:2834]: Good eye, thanks

**Comment by troygiorshev** [comment]
`git range-diff master 59df5d1 e85fb2d`

* Added missing lock annotations
* Cleaned PushMessage more thoroughly

File: src/net.cpp [hunk 2]
```diff
@@ -2813,35 +2813,37 @@ bool CConnman::NodeFullyConnected(const CNode* pnode)
 
 void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
 {
-    size_t nMessageSize = msg.data.size();
-    LogPrint(BCLog::NET, "sending %s (%d bytes) peer=%d\n",  SanitizeString(msg.m_type), nMessageSize, pnode->GetId());
+    size_t message_size = msg.data.size();
```
Comment (hebasto) [inline src/net.cpp:2814]: d093d261e2df27b2466ffe4e92bd48f22818d326
```suggestion
    const size_t message_size = msg.data.size();
```

but I think that the code would look better without the `message_size` variable at all.

File: src/net.cpp [hunk 3]
```diff
@@ -2813,35 +2813,37 @@ bool CConnman::NodeFullyConnected(const CNode* pnode)
...
+    LogPrint(BCLog::NET, "sending %s (%d bytes) peer=%d\n", SanitizeString(msg.m_type), message_size, pnode->GetId());
 
     // make sure we use the appropriate network transport format
-    std::vector<unsigned char> serializedHeader;
-    pnode->m_serializer->prepareForTransport(msg, serializedHeader);
-    size_t nTotalSize = nMessageSize + serializedHeader.size();
+    std::vector<unsigned char> serialized_header;
+    pnode->m_serializer->prepareForTransport(msg, serialized_header);
+    size_t total_size = message_size + serialized_header.size();
```
Comment (hebasto) [inline src/net.cpp:2820]: d093d261e2df27b2466ffe4e92bd48f22818d326
```suggestion
    const size_t total_size = message_size + serialized_header.size();
```

File: src/net.h [hunk 4]
```diff
@@ -817,7 +817,7 @@ class CNode
     std::atomic_bool fPauseSend{false};
 
 protected:
-    mapMsgCmdSize mapSendBytesPerMsgCmd;
+    mapMsgCmdSize mapSendBytesPerMsgCmd GUARDED_BY(cs_vSend);
```
Comment (hebasto) [inline src/net.h:820]: 9e92f63e2deadbfd7424ce4585f02ad580dcee8c

Is there any reason to keep `mapSendBytesPerMsgCmd` `protected`?

File: src/net.cpp [hunk 5]
```diff
@@ -1445,11 +1446,8 @@ void CConnman::SocketHandler()
         //
         if (sendSet)
         {
-            LOCK(pnode->cs_vSend);
             size_t nBytes = SocketSendData(pnode);
-            if (nBytes) {
-                RecordBytesSent(nBytes);
-            }
+            if (nBytes) RecordBytesSent(nBytes);
         }
```
Comment (hebasto) [inline src/net.cpp:1451]: e85fb2d34d8515513e404163cf149be61e1ed708

```
        if (sendSet) {
            const size_t bytes_sent = SocketSendData(pnode);
            if (bytes_sent) RecordBytesSent(bytes_sent);
        }
```

File: src/net.cpp [hunk 6]
```diff
@@ -2837,13 +2835,12 @@ void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
...
-        if (optimistic_send) {
-            bytes_sent = SocketSendData(pnode);
-        }
     }
-    if (bytes_sent) RecordBytesSent(bytes_sent);
+    // If write queue was empty, attempt "optimistic write"
+    if (optimistic_send == true) {
+        size_t bytes_sent{SocketSendData(pnode)};
+        if (bytes_sent) RecordBytesSent(bytes_sent);
+    }
```
Comment (hebasto) [inline src/net.cpp]: e85fb2d34d8515513e404163cf149be61e1ed708
```suggestion
    if (optimistic_send) {
        const size_t bytes_sent = SocketSendData(pnode);
        if (bytes_sent) RecordBytesSent(bytes_sent);
    }
```

**Review by hebasto — CHANGES_REQUESTED** [review]
Concept ACK.

The first commit message could be prefixed with `refactor: `, no?

File: src/net.cpp (hunk 2 above)
Comment (troygiorshev) [reply src/net.cpp:2814]: I don't have a strong opinion on the removal of `message_size` and/or `total_size`.  However such a change is a bit more than I'm looking to do here.

File: src/net.cpp (hunk 6 above)
Comment (troygiorshev) [reply src/net.cpp]: Thanks for catching the `const` on this one.  However, I prefer list initialization whenever possible, so I'm going to keep it unless you feel strongly against it.

File: src/net.h (hunk 4 above)
Comment (troygiorshev) [reply src/net.h:820]: Yes, it's to keep it consistent with `nSendBytes` for calls to the `getpeerinfo` rpc.

File: src/net.cpp (hunk 5 above)
Comment (troygiorshev) [reply src/net.cpp:1451]: Thanks.  Again I've preferred list initialization.

**Comment by troygiorshev** [comment]
`git range-diff master e85fb2d 9d47d6d`

* Clean commit improved
* Improved various const initializations
* Renamed the cleanup commit

**Comment by troygiorshev** [comment]
Seeing that #18784 is now closed, I've reworded the description of this PR.  This PR didn't end up being related to the bug #18784 was trying to fix, so this PR now stands on its own.

**Comment by troygiorshev** [comment]
Can't replicate CI failure. `mempool_updatefromblock.py` passes in 172 seconds on my machine.  (It was the longest test by a factor of 2x).  Didn't investigate further.

File: src/net.h (hunk 4 above)
Comment (jnewbery) [reply src/net.h:820]: There's no reason for either of these to be protected, since nothing inherits from `CNode`. However, that's separate from this PR and could be fixed elsewhere.

File: src/net.cpp (hunk 6 above)
Comment (jnewbery) [reply src/net.cpp]: I have a mild stylistic preference to use `=` when initializing a simple type with the return value of a function, and to use braced initialization for initializing from a literal. That's perhaps inconsistent, but it seems to be the style that most people gravitate towards. No strong preference since they both result in the same thing, but if it's a vote, I'd go for `=` :)

File: src/net.cpp [hunk 7]
```diff
@@ -2813,35 +2811,36 @@ bool CConnman::NodeFullyConnected(const CNode* pnode)
...
+    const size_t total_size{message_size + serialized_header.size()};
 
-    size_t nBytesSent = 0;
+    bool optimistic_send{false};
     {
         LOCK(pnode->cs_vSend);
-        bool optimisticSend(pnode->vSendMsg.empty());
+        optimistic_send = pnode->vSendMsg.empty();
 
         //log total amount of bytes per message type
```
Comment (jnewbery) [inline src/net.cpp:2827]: If you retouch the branch, you may as well touch _every_ line in this function:

```suggestion
        // Log total amount of bytes per message type
```

File: src/net.h [hunk 8]
```diff
@@ -759,8 +759,8 @@ class CNode
     // socket
     std::atomic<ServiceFlags> nServices{NODE_NONE};
     SOCKET hSocket GUARDED_BY(cs_hSocket);
-    size_t nSendSize{0}; // total size of all vSendMsg entries
-    size_t nSendOffset{0}; // offset inside the first vSendMsg already sent
+    size_t nSendSize GUARDED_BY(cs_vSend){0}; // total size of all vSendMsg entries
+    size_t nSendOffset GUARDED_BY(cs_vSend){0}; // offset inside the first vSendMsg already sent
     uint64_t nSendBytes GUARDED_BY(cs_vSend){0};
     std::deque<std::vector<unsigned char>> vSendMsg GUARDED_BY(cs_vSend);
     RecursiveMutex cs_vSend;
```
Comment (jnewbery) [inline src/net.h:766]: I think this could use a comment. Something like:

```
//* Guards the send buffer and statistics about data sent on this connection */
```

**Comment by jnewbery** [comment]
utACK 9d47d6da524b78ee4a628119f09636b855e881a2

Justification for the change in commit _Move cs_vSend lock into SocketSendData_ looks good to me.

I'd personally remove the easter egg from the _cleanup: Clean PushMessage_ commit log, but it's fine either way.

**Comment by troygiorshev** [comment]
Thanks for the review @jnewbery!  I'll leave your suggestions for now, and I'll look at adding them if there are non-stylistic changes needed for this PR.

My usage of list initialization comes almost completely from [this stackoverflow](https://stackoverflow.com/questions/18222926/why-is-list-initialization-using-curly-braces-better-than-the-alternatives), leaving the `X = Y` construct exclusively for assignment.  The C++ Core Guidelines talk about this as well, [ES.23](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es23-prefer-the--initializer-syntax)

**Comment by jnewbery** [comment]
> My usage of list initialization comes from this stackoverflow. The C++ Core Guidelines talk about this as well, ES.23.

> Edit: Another article about list initialization for those interested (link)

Thanks for the links. I retract my vote for = initialization!

File: src/net.h (hunk 8 above)
Comment (hebasto) [reply src/net.h:766]: The `GUARDED_BY` annotation seems quite self-documented; especially, when the mutex and all of the data members protected by it are nicely grouped :)

File: src/net.h [hunk 9]
```diff
@@ -759,8 +759,8 @@ class CNode
     // socket
     std::atomic<ServiceFlags> nServices{NODE_NONE};
     SOCKET hSocket GUARDED_BY(cs_hSocket);
-    size_t nSendSize{0}; // total size of all vSendMsg entries
-    size_t nSendOffset{0}; // offset inside the first vSendMsg already sent
+    size_t nSendSize GUARDED_BY(cs_vSend){0}; // total size of all vSendMsg entries
```
Comment (hebasto) [inline src/net.h:762]: 92897d43443db3a7cd68d8655d1bbf4b9b69c427

nit, clang-format suggests:
```suggestion
    size_t nSendSize GUARDED_BY(cs_vSend){0};   // total size of all vSendMsg entries
```

File: src/net.cpp [hunk 10]
```diff
@@ -747,8 +747,9 @@ void V1TransportSerializer::prepareForTransport(CSerializedNetMsg& msg, std::vec
     CVectorWriter{SER_NETWORK, INIT_PROTO_VERSION, header, 0, hdr};
 }
 
-size_t CConnman::SocketSendData(CNode *pnode) const EXCLUSIVE_LOCKS_REQUIRED(pnode->cs_vSend)
+size_t CConnman::SocketSendData(CNode *pnode) const
```
Comment (hebasto) [inline src/net.cpp:750]: 9d47d6da524b78ee4a628119f09636b855e881a2

nit, clang-format suggests:
```suggestion
size_t CConnman::SocketSendData(CNode* pnode) const
```

File: src/net.cpp [hunk 11]
```diff
@@ -1445,11 +1446,8 @@ void CConnman::SocketHandler()
         //
         if (sendSet)
         {
```
Comment (hebasto) [inline src/net.cpp:1448]: 9d47d6da524b78ee4a628119f09636b855e881a2, nit:
```suggestiom
        if (sendSet) {
```

**Review by hebasto — APPROVED** [review]
ACK 9d47d6da524b78ee4a628119f09636b855e881a2, I have reviewed the code and it looks OK, I agree it can be merged.

**Comment by troygiorshev** [comment]
Thanks for the ack hebasto!  I now have an ack from you and jnewbery.  Both of you gave style suggestions (which I agree with).  Since the suggestions are both further improvements to existing code (as opposed to style problems I'm adding myself, with the exception of the spaces before the comment) I'm going to leave them until a functional change is needed.

File: src/net.cpp [hunk 12]
```diff
@@ -2837,13 +2835,12 @@ void CConnman::PushMessage(CNode* pnode, CSerializedNetMsg&& msg)
...
-
-        // If write queue empty, attempt "optimistic write"
-        if (optimistic_send) {
-            bytes_sent = SocketSendData(pnode);
-        }
     }
-    if (bytes_sent) RecordBytesSent(bytes_sent);
+    // If write queue was empty, attempt "optimistic write"
+    if (optimistic_send) {
+        const size_t bytes_sent{SocketSendData(pnode)};
```
Comment (MarcoFalke) [inline src/net.cpp:2841]: Why is this change needed? Previously we'd optimistically write only one message, now we might write more than one message when other messages have been pushed by other threads.

**Review by maflcko — COMMENTED** [review]
What is the goal of this change? If it is to avoid a lock when `RecordBytesSent`, you can simply release the lock before calling `RecordBytesSent`. I don't think this justifies the other behaviour changes.

**Comment by jnewbery** [comment]
@MarcoFalke would you prefer an approach like https://github.com/jnewbery/bitcoin/tree/pr19673.1 ?

**Comment by MarcoFalke** [comment]
Yes, if the goal is to avoid a lock when RecordBytesSent, then simply releasing the lock before calling RecordBytesSent seems a good solution.

**Comment by jnewbery** [comment]
@MarcoFalke - I've opened #20816 as an alternative.

**Comment by troygiorshev** [comment]
I was too preoccupied with whether or not I could, I didn't stop to think whether or not I should :)

Closing this in favor of #20816.
